src/lib/formats/h17disk.cpp

```diff
@@ -0,0 +1,246 @@
+{
+	uint8_t h[4];
+	auto const [err, actual] = read_at(io, 0, h, 4);
+
+	if (err || (actual != 4))
+	{
+		return 0;
+	}
+
+	// Verify "H17D" Signature.
+	if ((h[0] == 0x48) && (h[1] == 0x31) && (h[2] == 0x37) && (h[3] == 0x44))
+	{
+		return FIFID_SIGN;
+	}
+
+	return 0;
+}
+
+bool heath_h17d_format::load(util::random_read &io, uint32_t form_factor, const std::vector<uint32_t> &variants, floppy_image &image) const
+{
+	const format fmt = find_format(io);
+
+	if (!fmt.head_count)
+	{
+		LOG_FORMATS("invalid format\n");
+
+		return false;
+	}
+
+	image.set_variant(fmt.variant);
+
+	std::vector<uint32_t> buf;
+
+	auto const [secm_pos, secm_length] = find_block(io, SecM);
+	
+	uint8_t sector_meta_data[SECTOR_METADATA_SIZE];
+	uint8_t sector_data[SECTOR_DATA_SIZE];
+
+	for (int head = 0; head < fmt.head_count; head++)
+	{
+		for (int track = 0; track < fmt.track_count; track++)
+		{
+			for (int sector = 0; sector < SECTORS_PER_TRACK; sector++)
+			{
+				int sect_meta_pos = (sector + (track * fmt.head_count + head) * SECTORS_PER_TRACK) * SECTOR_METADATA_SIZE + secm_pos;
+
+				auto const [err, actual] = read_at(io, sect_meta_pos, sector_meta_data, SECTOR_METADATA_SIZE);
+
+				if (err || (actual != SECTOR_METADATA_SIZE))
+				{
+					LOG_FORMATS("unable to read sect meta data %d/%d/%d\n", head, track, sector);
+
+					return false;
+				}
+				int data_offset = sector_meta_data[0] << 24 | sector_meta_data[1] << 16 | sector_meta_data[2] << 8 | sector_meta_data[3];
+
+				auto const [err2, actual2] = read_at(io, data_offset, sector_data, SECTOR_DATA_SIZE);
+
+				if (err2 || (actual2 != SECTOR_DATA_SIZE))
+				{
+					LOG_FORMATS("unable to read sect data %d/%d/%d\n", head, track, sector);
+
+					return false;
+				}
+
+				// Inital 15 zero bytes
+				for (int i = 0; i < 15; i++)
+				{
+					fm_reverse_byte_w(buf, 0);
+				}
+
+				// header (sync byte, volume, track, sector, checksum)
+				for (int i = 0; i < 5; i++)
+				{
+					fm_reverse_byte_w(buf, sector_meta_data[5 + i]);
+				}
+
+				// 12 zero bytes
+				for (int i = 0; i < 12; i++)
+				{
+					fm_reverse_byte_w(buf, 0);
+				}
+
+				// data sync byte
+				fm_reverse_byte_w(buf, sector_meta_data[10]);
+
+				// sector data
+				for (int i = 0; i < 256; i++)
+				{
+					fm_reverse_byte_w(buf, sector_data[i]);
+				}
+
+				// sector data checksum
+				fm_reverse_byte_w(buf, sector_meta_data[11]);
+
+				// trailing zero's until the next sector hole usually ~ 30 characters.
+				while (buf.size() < TRACK_SIZE / SECTORS_PER_TRACK * (sector + 1))
+				{
+					fm_reverse_byte_w(buf, 0);
+				}
+			}
+
+			generate_track_from_levels(track, head, buf, 0, image);
+			buf.clear();
+		}
+	}
+
+	return true;
+}
+
+void heath_h17d_format::fm_reverse_byte_w(std::vector<uint32_t> &buffer, uint8_t val) const
+{
+	static unsigned char lookup[16] = { 0x0, 0x8, 0x4, 0xc, 0x2, 0xa, 0x6, 0xe, 0x1, 0x9, 0x5, 0xd, 0x3, 0xb, 0x7, 0xf };
```

Comment:
Non-const static data is a bug waiting to happen…

Reply:
Thanks, I'll create a PR to address this.

Reply:
Oh, I see you already cleaned it up, thanks!